src/v/cloud_topics/level_zero/stm/ctp_stm_api.cc

```diff
@@ -29,93 +34,98 @@ std::ostream& operator<<(std::ostream& o, ctp_stm_api_errc errc) {
         return o << "timeout";
     case ctp_stm_api_errc::not_leader:
         return o << "not_leader";
+    case ctp_stm_api_errc::shutdown:
+        return o << "shutdown";
+    case ctp_stm_api_errc::failure:
+        return o << "failure";
     }
 }
 
-ctp_stm_api::ctp_stm_api(ss::logger& logger, ss::shared_ptr<ctp_stm> stm)
-  : _logger(logger)
+ctp_stm_api::ctp_stm_api(retry_chain_node& rtc, ss::shared_ptr<ctp_stm> stm)
+  : _rtc(&rtc)
+  , _rtclog(cd_log, _rtc, stm->ntp().path())
   , _stm(std::move(stm)) {}
 
-ss::future<> ctp_stm_api::stop() { co_await _gate.close(); }
-
-ss::future<checked<dl_snapshot_id, ctp_stm_api_errc>>
-ctp_stm_api::start_snapshot() {
-    vlog(_logger.debug, "Replicating ctp_stm_cmd::start_snapshot_cmd");
-    auto h = _gate.hold();
-
-    storage::record_batch_builder builder(
-      model::record_batch_type::ctp_stm_command, model::offset(0));
-    builder.add_raw_kv(
-      serde::to_iobuf(ctp_stm_key::start_snapshot),
-      serde::to_iobuf(start_snapshot_cmd()));
+ss::future<std::expected<model::offset, ctp_stm_api_errc>>
+ctp_stm_api::replicated_apply(model::record_batch&& batch) {
+    model::term_id term = _stm->_raft->term();
 
-    auto batch = std::move(builder).build();
+    vlog(
+      _rtclog.debug, "Replicating batch {} in term {}", batch.header(), term);
 
-    auto apply_result = co_await replicated_apply(std::move(batch));
-    if (apply_result.has_failure()) {
-        co_return apply_result.error();
-    }
+    auto opts = raft::replicate_options(raft::consistency_level::quorum_ack);
+    opts.set_force_flush();
+    auto res = co_await _stm->_raft->replicate(term, std::move(batch), opts);
 
-    // We abuse knowledge of implementation detail here to construct the
-    // dl_snapshot_id without having to setup listeners and notifiers of command
-    // apply.
-    auto expected_id = dl_snapshot_id(dl_version(apply_result.value()));
-
-    // Ensure that the expected snapshot was created.
```

Comment:
please consider the impact of code placement on diff construction. if you put new code far away in the file (e.g. at the end) then diff will be more likely to not interleave the hunks and you'll get bigger sections of contiguous `+` and `-` lines.